Crypto/ECC/Ed25519.hs

```diff
@@ -0,0 +1,298 @@
+-- Maintainer  : Olivier Chéron <olivier.cheron@gmail.com>
+-- Stability   : experimental
+-- Portability : unknown
+--
+-- Ed25519 arithmetic primitives.
+--
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+module Crypto.ECC.Ed25519
+    ( Scalar
+    , Point
+    -- * Scalars
+    , scalarGenerate
+    , scalarDecodeLong
+    , scalarEncode
+    -- * Points
+    , pointDecode
+    , pointEncode
+    -- * Arithmetic functions
+    , toPoint
+    , scalarAdd
+    , scalarMul
+    , pointNegate
+    , pointAdd
+    , pointDouble
+    , pointMul
+    , pointsMulVarTime
+    ) where
+
+import           Data.Bits
+import           Data.Word
+import           Foreign.C.Types
+import           Foreign.Ptr
+import           Foreign.Storable
+
+import           Crypto.Error
+import           Crypto.Internal.ByteArray (ByteArrayAccess, Bytes,
+                                            ScrubbedBytes, withByteArray)
+import qualified Crypto.Internal.ByteArray as B
+import           Crypto.Internal.Compat
+import           Crypto.Internal.Imports
+import           Crypto.Random
+
+
+scalarArraySize :: Int
+scalarArraySize = 40 -- maximum [9 * 4 {- 32 bits -}, 5 * 8 {- 64 bits -}]
+
+-- | A scalar modulo order of curve Ed25519.
+newtype Scalar = Scalar ScrubbedBytes
+    deriving (Show,NFData)
+
+instance Eq Scalar where
+    (Scalar s1) == (Scalar s2) = unsafeDoIO $
+        withByteArray s1 $ \ps1 ->
+        withByteArray s2 $ \ps2 ->
+            fmap (/= 0) (ed25519_scalar_eq ps1 ps2)
+    {-# NOINLINE (==) #-}
+
+pointArraySize :: Int
+pointArraySize = 160 -- maximum [4 * 10 * 4 {- 32 bits -}, 4 * 5 * 8 {- 64 bits -}]
+
+-- | A point on curve Ed25519.
+newtype Point = Point Bytes
+    deriving NFData
+
+instance Show Point where
+    showsPrec d p =
+        let bs = pointEncode p :: Bytes
+         in showParen (d > 10) $ showString "Point "
+                               . shows (B.convertToBase B.Base16 bs :: Bytes)
+
+instance Eq Point where
+    (Point p1) == (Point p2) = unsafeDoIO $
+        withByteArray p1 $ \pp1 ->
+        withByteArray p2 $ \pp2 ->
+            fmap (/= 0) (ed25519_point_eq pp1 pp2)
+    {-# NOINLINE (==) #-}
+
+-- | Generate a random scalar.
+scalarGenerate :: MonadRandom randomly => randomly Scalar
+scalarGenerate = unwrap . scalarDecodeLong . clamp <$> generate
+  where
+    unwrap (CryptoPassed x) = x
+    unwrap (CryptoFailed _) = error "scalarGenerate: assumption failed"
+
+    generate :: MonadRandom randomly => randomly ScrubbedBytes
+    generate = getRandomBytes 32
+
+    -- Uses the same bit mask than during key-generation procedure,
+    -- but without making divisible by 8.  As a consequence of modular
+    -- reduction, distribution is not uniform.  But the curve order is
+    -- very close to 2^252 so only a tiny fraction of the scalars have
+    -- lower probability, roughly 1/(2^126) of all possible values.
+    clamp :: ByteArrayAccess ba => ba -> ScrubbedBytes
+    clamp bs = B.copyAndFreeze bs $ \p -> do
```

Comment:
I'm not sure about this; so there's no reason to make it divisible by 8 for ed25519, however with the birationally equivalence with curve25519 and curve25519 low order requirements (divisible by 8), I'm not sure how to best handle this in a non-confusing way or if we should at all.

The same thing with the rest of the clamping, the reason for it to be done is actually slightly obscure and not as specified as I would like (I spent a good amount of time of this very problem recently). The only reason that I found for this clamping is to actually prevent a timing issue with multiplication overflowing, but once you take away the specific usage to the signature system (ed25519), and instead do generic mathematical operation on `edwards25519` does it make sense to still have this ?

Reply:
The masking of highest bits is something I tried to understand too and indeed is not really explained. In addition to having a fixed magnitude for timing reasons, I wanted to limit the number of times values wrap around the curve order and decrease the fraction of scalars with statistical bias. However I see libraries take another approach, actually using many more bits [like this](https://github.com/warner/python-spake2/blob/05b9f968d37dc5419f0e6e20c9b65737de21a7e9/src/spake2/ed25519_basic.py#L162) so I think I'm going to change that part.

The scalar functions work modulo prime order L, not 8*L, so masking the lower bits does not provide divisibility by 8. Here the scalars will be most useful when applied to points in the prime-order subgroup.

It's true all of this gives enough power to shoot oneself in the foot with the cofactor.
However I don't see how to keep generality to implement many scheme and at the same time prevent any effect of cofactor (besides the Decaf technique of switching to another curve).

But clearly this deserves some warnings in documentation.